### DISCUSSION TITLE
How to redirect back to previous URL from API route when redirecting to API route

### DISCUSSION THREAD

**Discussion by niels-bosman** [vercel/next.js#63723] [Help]
### Summary

We run the following code for our users when they do not have a unique session token yet. This route should return the user to the original URL after successful cookie setting.

`/app/api/session/route.ts`
```ts
// Assign a new session token to the user and redirect them to the home page again.
export const GET = async ()  => {
  if (!cookies().has(SESSION_TOKEN_IDENTIFIER)) {
    cookies().set(SESSION_TOKEN_IDENTIFIER, createSessionToken(), {
      maxAge: SESSION_TOKEN_MAX_AGE,
    });
  }

  return redirect('/'); // Should go back to last URL
};
```

The problem here is that we have no referrer on the request, because the user comes to this route via a redirect in `RootLayout`.

in `app/layout.tsx`:

```ts
...
const sessionToken = getSessionToken();
if (!sessionToken) redirect('/api/session', RedirectType.push);
```

What is the recommended way to redirect BACK to the previous URL where the initial redirect happens from? I don’t want to make it so users are always redirected back to "/".

**Comment by don-esteban** [comment]
Would it do the job, if you [read the referer](https://nextjs.org/docs/app/api-reference/functions/headers) in your Layout component and pass it somehow (perhaps as [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams)) in the redirect?

**Reply by niels-bosman** [reply]
@don-esteban Thanks for your suggestion! The problem is, when a user comes to the page for the first time the referer will be `null`. So there is no way to get the URL of the server component inside of it (as far as I know).

**Reply by don-esteban** [reply]
Perhaps I do not understand exactly what you want to achieve.

1. Your user hits your page the first time from an external site: It depends on the external Site. Not all sites/links do send referers. Probably not what you do.
2. Your user hits your page the first time from an internal link: Use `<a>` and not the Next.js `<Link>` component.

For me, it works as expected.

Not sure about caching with the `<Link>` component. This can perhaps be solved with the right caching strategy.

**Reply by niels-bosman** [reply]
@don-esteban Ah I think I might have not made the intentions clear enough.

First time users might go to our `/thank-you` page (redirected from a Stripe payment). This page is wrapped by my general `app/layout.tsx` which checks if a user has a session token (via redirecting to `/api/session` like seen in my example).

From this `/api/session/route.ts` ideally I would send the users back to `/thank-you` after the cookie was successfully set, but because it was a redirect and the user never actually been on the page, this is not set as the referer.

Does this explanation make sense?

**Reply by don-esteban** [reply]
If I think step by step

1. User hits `/thank-you` page coming from somewhere (Stripe/external link)
2. Page responds with redirect (307) 

Referer won't help you. Even if Next.js `redirect` preserves the referer, it would be the external page. But what about my proposal passing the `/thank-you` slug as URLSearchParams to the API route?

`if (!sessionToken) redirect('/api/session?callback-path=/thank-you', RedirectType.push);`

If you don't know the callback path `/thank-you` ahead of time, perhaps [dynamic routes](https://nextjs.org/docs/app/api-reference/file-conventions/layout#params-optional) can help to get a slug.

**Reply by niels-bosman** [reply]
@don-esteban Yeah, right. I think that would be the best solution, but as far as I know it is not possible to get the current route inside of a server component root layout which seems to be the bottleneck for me :(...

**Reply by don-esteban** [reply]
Yes, this can be tricky, if dynamic routes don't help. It depends on the architecture of your app.

**Comment by don-esteban** [comment]
I'm not sure, because I still had no similar use case to solve, but I would first examine how middleware can handle this.

Intercept requests in middleware, examine resource paths and session cookies and decide to which page (redirect, rewrite) to go. Please tell me if this works. Soon, I'll be facing similar requirements.

**Reply by niels-bosman** [reply]
Yeah middleware would probably be the best solution here since it seems to be intended that the layout does not know the URL (because it never re-renders on page change). Unfortunately it is not really feasible to enable middleware for us for every URL since we get a lot of traffic and middleware invocations can be expensive when hosting on Vercel...

See: https://github.com/vercel/next.js/issues/43704

Thanks for helping out.

**Comment by don-esteban** [comment]
Do not dismiss middleware based on arbitrary comments. I would go for a clean and simple solution and optimize, if I hit a bottleneck. Not sure, if all comments are factbased. https://nextjs.org/docs/app/building-your-application/routing/middleware. Good luck!

**Reply by niels-bosman** [reply]
Thanks! Probably gonna try a client side redirect solution to make sure the referer is set :)
